### DISCUSSION TITLE
CUPS does not detect Canon SELPHY CP1500 via USB

### DISCUSSION THREAD

**Discussion by manfred-raab** [OpenPrinting/cups#994] [Q&A] [answered]
Hello,

it looks like there are a lot of printing experts here. I am trying to get a Canon SELPHY CP1500 to work via USB. Raspberry Pi 5 with Bookworm 64-bit OS.

$ lsusb:

Bus 001 Device 004: ID 04a9:3302 Canon, Inc. SELPHY CP1500

$ sudo usb-devices:

T:  Bus=01 Lev=01 Prnt=01 Port=00 Cnt=01 Dev#=  4 Spd=480 MxCh= 0
D:  Ver= 2.00 Cls=00(>ifc ) Sub=00 Prot=00 MxPS=64 #Cfgs=  1
P:  Vendor=04a9 ProdID=3302 Rev=00.01
S:  Manufacturer=Canon
S:  Product=SELPHY CP1500
S:  SerialNumber=B122080623375378
C:  #Ifs= 2 Cfg#= 1 Atr=c0 MxPwr=0mA
I:  If#= 0 Alt= 1 #EPs= 2 Cls=07(print) Sub=01 Prot=04 Driver=usbfs
E:  Ad=01(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=82(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:  If#= 1 Alt= 1 #EPs= 2 Cls=07(print) Sub=01 Prot=04 Driver=usbfs
E:  Ad=05(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=86(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms

However CUPS does not detect this Printer and I can not add it.

Does anybody can give me a hint whats wrong and whats to do. 

I'm desperate.

Thanks in advance

Manfred

**Comment by szlt5** [comment, marked as answer]
Check if ipp-usb is running. If ipp-usb is running, ipp-usb will always occupy the usb interface.
`$ps -ef | grep ipp-usb`

Or try turning off the Airprint function on the printer and connecting the USB again.

**Reply by manfred-raab** [reply]
Thank you very much. I have uninstalled ipp-usb and indeed CUPS has now shown me the printer. I have now installed the printer with a PPD file and Printer is shown as idle. Unfortunately, a test print has not yet worked. 

Job Page: 
ID Name User Size Pages State Control
Canon_SELPHY_CP1500-5 Unknown Withheld 5633k Unknown pending since
Wed 26 Jun 2024 15:34:43 CEST 

The print job seems to be waiting for the printer but it does not seem to be responding. I'll have to analyse this further.

**Reply by zdohnal** [reply]
>If ipp-usb is running, ipp-usb will always occupy the usb interface.

@szlt5 This is not entirely true - it does occupy usb interface if the device has an IPP over USB interface, it does not occupy USB port of all USB printers around  - the "old" (approx. 2012 or older) devices are not affected.

@manfred-raab if the device has AirPrint functionality (check in settings on the printer panel) enabled and IPP-over-USB interface exists, then the device should work via ipp-usb - if it does not, then there can be issue in the device, in ipp-usb, in mDNS, or in CUPS. I see old issue on ipp-usb - https://github.com/OpenPrinting/ipp-usb/issues/73 - and the issue with Selphy CP1500 looks fixed. Your system probably runs some older version of ipp-usb.

If you opt-out from ipp-usb, you will have to figure out the driver what the printer needs if you don't have the exact one.

**Reply by szlt5** [reply]
After searching, it seems that the manufacturer does not provide a specific print driver (or specific ppd file) for the SELPHY CP1500 on Linux. 
So it is still recommended to print with ipp-usb + driverless or ipp-usb + everywhere.

**Reply by manfred-raab** [reply]
Thank you for sharing your Experiences. I have now created a fresh Bookworm image. Ipp-usb is running and the Canon quirk is also present. In Cups the printer is shown as a local printer. But shouldn't it appear as a Discovered Network Printer through ipp-usb?

**Reply by speachy** [reply]
Just as an FYI -- For the SELPHY printers to work over IPP-USB, the printer has to have a working wifi connection.

(There's also a firmware update for the CP1500 that fixes some IPP related issues)

**Reply by zdohnal** [reply]
@manfred-raab If you have driverless backend installed from cups-filters project, IMO the device should be among Discovered Network printers in the web ui (if that's what you mean by the name). The backend finds the service on mDNS and presents it on request from cupsd when finding devices, however it is not required for device to work.

ipp-usb registers the printer as IPP service on localhost, where cupsd is able to pick it up as a temporary queue - the queue is constructed when you try to print and removed after successful printing. Most recent applications (PDF viewer aka evince, Firefox or anything else based on GTK3 and Qt) are able to use them, so printer doesn't have to be installed by user. AFAIK the problematic ones Google Chrome and Chromium, where the dialog doesn't use them.

**Reply by weidi** [reply]
As i was using this Issue as help i may add my two cents.
I tried this on debian but they use ipp-usb 0.9.23 and there is a fix in 0.9.24 so i installed 0.9.28 from ubuntu that helped a bit.
What finally solved my issues was using fedora, the rely on ipp-usb 0.9.30 currently and it really is a plug and play situation now.
